{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b83ce32c",
   "metadata": {},
   "source": [
    "### Practical 01\n",
    "\n",
    "#### Task 1\n",
    "\n",
    "Consider a sample space consisting of all possible amino acids. Do the sets of amino acids with __polar uncharged side chain__ and amino acids __with G as first base__ form a partition of the sample space? Use __R__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a92e972",
   "metadata": {},
   "outputs": [],
   "source": [
    "# amino acids\n",
    "S <- c(\"F\", \"L\", \"V\", \"M\", \"I\", \"S\", \"P\", \"T\", \"A\", \"Y\", \"H\", \"Q\", \"N\", \"K\", \"D\", \"E\", \"C\", \"W\", \"R\", \"G\")\n",
    "\n",
    "# amino acids with polar uncharged side chain\n",
    "A <- c(\"S\", \"T\", \"N\", \"Q\")\n",
    "\n",
    "# amino acids with G as first base\n",
    "B <- c(\"V\", \"A\", \"D\", \"E\", \"G\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa53e78e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1) are they disjoint?\n",
    "length(intersect(A, B))==0\n",
    "\n",
    "# 2) do they form a partition?\n",
    "length(setdiff(S, union(A, B)))==0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5638b8ab",
   "metadata": {},
   "source": [
    "#### Task 2\n",
    "\n",
    "Given the sample space and events defined in Task 1, use __R__ to prove the set properties of commutativity and associativity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c643133c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# commutativity\n",
    "setequal(union(A,B), union(B,A))\n",
    "\n",
    "setequal(intersect(A,B), intersect(B,A))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45088b38",
   "metadata": {},
   "outputs": [],
   "source": [
    "C <- c()\n",
    "\n",
    "# associativity\n",
    "setequal(union(A,union(B,C)), union(union(A,B),C))\n",
    "\n",
    "setequal(intersect(A,intersect(B,C)), intersect(intersect(A,B),C))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6c227f1",
   "metadata": {},
   "source": [
    "#### Task 3\n",
    "\n",
    "The canonical protein sequence for Collagen alpha-1(I) chain is available in the `Data` folder as `P02452.fasta` file and has been obtained from `https://rest.uniprot.org/uniprotkb/P02452.fasta`.\n",
    "The sequence is in `fasta` format which includes a header in the first line.\n",
    "\n",
    "Using __R__, calculate the probability of drawing one amino acid with polar uncharged side chain from the whole sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbc28183",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein <- readLines(con = \"Data/P02452.fasta\")\n",
    "\n",
    "protein <- paste(protein[-1], sep=\"\", collapse=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5c86664",
   "metadata": {},
   "outputs": [],
   "source": [
    "nr_A <- length(which(strsplit(protein, split=\"\")[[1]] %in%A ))\n",
    "nr_S <- nchar(protein)\n",
    "nr_A / nr_S"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15feffbf",
   "metadata": {},
   "source": [
    "#### Task 4\n",
    "\n",
    "Perform a sliding-window analysis on the protein by recording the occurrences (of frequencies) of amino acid with polar uncharged side chain over the Collagen protein sequence.\n",
    "\n",
    "In a sliding-window analysis, you are requested to scan the sequence over windows of fixed length and record (and plot) the metric of interest. Windows can be overlapping or not.\n",
    "\n",
    "You are free to decide on the windows' length and step (how much the move from one window to the next one), the latter being relevant only for overlapping windows. It is easier to implement non-overlapping windows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d356c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "len_protein <- nchar(protein)\n",
    "\n",
    "len_win <- 100\n",
    "\n",
    "starts <- seq(1, len_protein, len_win)\n",
    "\n",
    "starts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9975a7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "values <- c()\n",
    "\n",
    "for (start in starts) {\n",
    "    \n",
    "    end <- start + len_win -1\n",
    "    \n",
    "    window <- strsplit(protein, split=\"\")[[1]][start:end]\n",
    "    \n",
    "    count <- length(which(window %in%A ))\n",
    "      \n",
    "    len_window <- length(which(!is.na(window)))\n",
    "    \n",
    "    values <- c(values, count / len_window)\n",
    "    \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e3dd042",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot(values, xlab=\"Window\", type=\"l\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
